chain/network/src/accounts_data/mod.rs

```diff
@@ -0,0 +1,204 @@
+//! - we can afford verifying a few invalid signatures per SyncAccountsData message.
+//!
+//! Strategy:
+//! - handling of SyncAccountsData should be throttled by PeerActor/PeerManagerActor.
+//! - synchronously select interesting AccountData (i.e. those with never timestamp than any
+//!   previously seen for the given (account_id,epoch_id) pair.
+//! - asynchronously verify signatures, until an invalid signature is encountered.
+//! - if any signature is invalid, drop validation of the remaining signature and ban the peer
+//! - all valid signatures verified, so far should be inserted, since otherwise we are open to the
+//!   following attack:
+//!     - a bad peer may spam us with <N valid AccountData> + <1 invalid AccountData>
+//!     - we would validate everything every time, realizing that the last one is invalid, then
+//!       discarding the progress
+//!     - banning a peer wouldn't help since peers are anonymous, so a single attacker can act as a
+//!       lot of peers
+use crate::network_protocol;
+use crate::network_protocol::SignedAccountData;
+use near_network_primitives::types::{AccountKeys, WithHash};
+use near_primitives::types::{AccountId, EpochId};
+use parking_lot::RwLock;
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+
+#[cfg(test)]
+mod tests;
+
+/// Runtime is a wrapper of tokio::runtime::Runtime
+/// which starts rutime shutdown when dropped.
+// TODO(gprusak): replace Option with ManuallyDrop?
+struct Runtime(Option<tokio::runtime::Runtime>);
+
+impl std::ops::Deref for Runtime {
+    type Target = tokio::runtime::Runtime;
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref().unwrap()
+    }
+}
+
+impl Drop for Runtime {
+    fn drop(&mut self) {
+        self.0.take().unwrap().shutdown_background()
+    }
+}
+
+impl Runtime {
+    fn new() -> Self {
+        Self(Some(tokio::runtime::Runtime::new().unwrap()))
+    }
+}
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub(crate) enum Error {
+    #[error("found an invalid signature")]
+    InvalidSignature,
+    #[error("found too large payload")]
+    DataTooLarge,
+    #[error("found multiple entries for the same (epoch_id,account_id)")]
+    SingleAccountMultipleData,
+}
+
+struct CacheInner {
+    keys: Arc<AccountKeys>,
+    /// Current state of knowledge about an account.
+    /// key is the public key of the account in the given epoch.
+    /// It will be used to verify new incoming versions of SignedAccountData
+    /// for this account.
+    data: HashMap<(EpochId, AccountId), SignedAccountData>,
+}
+
+impl CacheInner {
+    fn is_new(&self, d: &SignedAccountData) -> bool {
+        let id = (d.epoch_id.clone(), d.account_id.clone());
+        self.keys.contains_key(&id)
+            && match self.data.get(&id) {
+                Some(old) if old.timestamp >= d.timestamp => false,
+                _ => true,
+            }
+    }
+    fn try_insert(&mut self, d: SignedAccountData) -> Option<SignedAccountData> {
```

Comment:
I wonder if there's some lightweight way to distinguish potentially unverifid data in the first-case from the verified data in the second case. Probably isn't worth it though.

Reply:
I'm afraid I do not follow your concern here.